ciao-sdk/show.go

```diff
@@ -0,0 +1,459 @@
+	} else {
+		url = buildComputeURL("%s/events", tenant)
+	}
+
+	resp, err := sendHTTPRequest("GET", url, nil, nil)
+	if err != nil {
+		fatalf(err.Error())
+	}
+
+	err = unmarshalHTTPResponse(resp, &events)
+	if err != nil {
+		fatalf(err.Error())
+	}
+
+	if Template != "" {
+		return tfortools.OutputToTemplate(os.Stdout, "event-list", Template,
+			&events.Events, nil)
+	}
+
+	fmt.Printf("%d Ciao event(s):\n", len(events.Events))
+	for i, event := range events.Events {
+		fmt.Printf("\t[%d] %v: %s:%s (Tenant %s)\n", i+1, event.Timestamp, event.EventType, event.Message, event.TenantID)
+	}
+	return nil
+}
+
+func dumpImage(i *image.DefaultResponse) {
+	fmt.Printf("\tName             [%s]\n", *i.Name)
+	fmt.Printf("\tSize             [%d bytes]\n", i.Size)
+	fmt.Printf("\tUUID             [%s]\n", i.ID)
+	fmt.Printf("\tStatus           [%s]\n", i.Status)
+	fmt.Printf("\tVisibility       [%s]\n", i.Visibility)
+	fmt.Printf("\tTags             %v\n", i.Tags)
+	fmt.Printf("\tCreatedAt        [%s]\n", i.CreatedAt)
+}
+
+func listImage(cmd *cobra.Command, args []string) error {
+	var t *template.Template
+	var err error
+	if Template != "" {
+		t, err = tfortools.CreateTemplate("image-list", Template, nil)
+		if err != nil {
+			fatalf(err.Error())
+		}
+	}
+
+	url := buildImageURL("images")
+	resp, err := sendHTTPRequest("GET", url, nil, nil)
+	if err != nil {
+		fatalf(err.Error())
+	}
+	defer func() { _ = resp.Body.Close() }()
+
+	if resp.StatusCode != http.StatusOK {
+		fatalf("Image list failed: %s", resp.Status)
+	}
+
+	var images = struct {
+		Images []image.DefaultResponse
+	}{}
+
+	err = unmarshalHTTPResponse(resp, &images)
+	if err != nil {
+		fatalf(err.Error())
+	}
+
+	if t != nil {
+		if err = t.Execute(os.Stdout, &images.Images); err != nil {
+			fatalf(err.Error())
+		}
+		return nil
+	}
+
+	for k, i := range images.Images {
+		fmt.Printf("Image #%d\n", k+1)
+		dumpImage(&i)
+		fmt.Printf("\n")
+	}
+
+	return err
+}
+
+func dumpInstance(server *compute.ServerDetails) {
+	fmt.Printf("\tUUID: %s\n", server.ID)
+	fmt.Printf("\tStatus: %s\n", server.Status)
+	fmt.Printf("\tPrivate IP: %s\n", server.PrivateAddresses[0].Addr)
+	fmt.Printf("\tMAC Address: %s\n", server.PrivateAddresses[0].MacAddr)
+	fmt.Printf("\tCN UUID: %s\n", server.NodeID)
+	fmt.Printf("\tTenant UUID: %s\n", server.TenantID)
+	if server.SSHIP != "" {
+		fmt.Printf("\tSSH IP: %s\n", server.SSHIP)
+		fmt.Printf("\tSSH Port: %d\n", server.SSHPort)
+	}
+
+	for _, vol := range server.Volumes {
+		fmt.Printf("\tVolume: %s\n", vol)
+	}
+}
+
+func listNodeInstances(node string) error {
+	if node == "" {
+		fatalf("Missing required -cn parameter")
+	}
+
+	var servers types.CiaoServersStats
+	url := buildComputeURL("nodes/%s/servers/detail", node)
+
+	resp, err := sendHTTPRequest("GET", url, nil, nil)
+	if err != nil {
+		fatalf(err.Error())
+	}
+
+	err = unmarshalHTTPResponse(resp, &servers)
+	if err != nil {
+		fatalf(err.Error())
+	}
+
+	for i, server := range servers.Servers {
```

Comment:
One thing @rbradford and I have discussed in the past is using tfortools to do all the formatting.

You could replace this whole loop with

tfortools.OutputToTemplate(os.Stdout, "servers", '{{table .}}', nil)

It would also make the code easier for all our list and show functons as there would only be one code path.  If the template string is not defined it would default to a template that displays things in the default format, otherwise you use the user supplied template.  But in all cases you output via a template.

Reply:
Actually, why is there no template support here?